### ISSUE TITLE
Releases on GitHub

### ISSUE THREAD

**Issue by brunonaibert** [the-tcpdump-group/tcpslice#7]
I am doing quality assurance work on the tcpslice package in Debian.
I updated the package's homepage to tcpdump.org and would like to change the check for new versions.
Currently the check is performed at ftp://ftp.ee.lbl.gov/ I would like to switch to GitHub.
Could you insert version 1.2a3 on GitHub and create a release of this code?
If possible, it would be interesting to have all possible versions ported on GitHub.
And the use of releases so that we can be alerted in Debian about new versions of this application.

**Comment by guyharris** [comment]
> I am doing quality assurance work on the tcpslice package in Debian.
> I updated the package's homepage to tcpdump.org and would like to change the check for new versions.
> Currently the check is performed at ftp://ftp.ee.lbl.gov/ I would like to switch to GitHub.
> Could you insert version 1.2a3 on GitHub

What do you mean by "insert version 1.2a3 on GitHub"?  "Add a tag for 1.2a3"?

> and create a release of this code?

"Create a release" in what sense?  "Add tcpslice to the www.tcpdump.org home page, including a downloadable 1.2a3 release tarball"?

We don't host libpcap or tcpdump releases on GitHub, we host them on www.tcpdump.org.  I don't see why tcpslice should be any different here.

Note also that the code on the master branch on GitHub has a number of changes made subsequent to the last LBL release; we may not have much interest in making a 1.2a3 release tarball, just as we don't have libpcap 0.4 and tcpdump 3.4 release tarballs in https://www.tcpdump.org/release/.

> If possible, it would be interesting to have all possible versions ported on GitHub.

If we make a tcpslice release in the future, we'd presumably create a branch for it, e.g. tcpslice-1.3, and create tags for 1.3.0, 1.3.1, etc..  We'd still host tcpslice-1.3.tar.gz on https://www.tcpdump.org/release/.

> And the use of releases so that we can be alerted in Debian about new versions of this application.

I'm not sure how new libpcap and tcpdump releases are announced; I didn't see any emails to tcpdump-workers or tcpdump-announce for tcpdump 4.9.3, for example.  @mcr, how are new tcpdump and libpcap releases announced?

**Comment by infrastation** [comment]
For tcpslice I had it in my mind to look through LBL sources and CVS-inherited branches to see how they relate to things in the git repository (i.e. no relevant bugfixes left behind), if that helps I can post results here if that turns out to be a quick check.

Regarding GitHub "releases", that's a web-site feature around git tags, which ("releases", not tags) I agree would muddy the water compared to plain old tarballs on tcpdump.org. Let's have proper tags and proper tarballs, a few automatic systems already use that to detect new tcpdump releases.

Considering the very low average volume of commits in tcpslice, I would not bother introducing (thus maintaining) one more "stable" branch and would just keep the master branch in a good shape (which is now easier with CI in place).

Additional input is welcome, of course.

**Comment by mcr** [comment]
There should have been announcements on tcpdump-announce for 4.9.3, if that didn't happen then mea-culpa.  We haven't actively maintained tcpslice much.  We mirrored it into github in the hope that it would be easier for contributors to do things.  If there is someone at debian who wants to merge code, that would be rather great.
(I don't think there is anything in the LBL sources or CVS which we have left behind)
[Maybe there are patches in the FreeBSD or NetBSD src tree]
Who is using tcpslice?  Often it's the wrong tool for the job, but the only tool available.

**Comment by fenner** [comment]
> Often it's the wrong tool for the job

Is there a better tool for combining multiple dumps into a single pcap file?

I think Denis is right that adding a tag to master to indicate a release is enough.  And I think it's reasonable to call *something* version 1.3.

**Comment by brunonaibert** [comment]
I have little experience with GitHub and tcpslice, sorry for the confusion.
I saw that they were active in the development and that we still don't have a tag indicating a release.
I will accompany the repository for when we have to introduce this version to Debian.
Thank you for your attention and clarification.

**Comment by mcr** [comment]
> > Often it's the wrong tool for the job
> 
> Is there a better tool for combining multiple dumps into a single pcap file?

No, but as I recall, it also lets you extract pieces of a pcap file by date, and usually that works out poorly.  There is no better tool :-)

> I think Denis is right that adding a tag to master to indicate a release is enough. And I think it's reasonable to call _something_ version 1.3.

Agreed

**Comment by guyharris** [comment]
> > > Often it's the wrong tool for the job
> > 
> > 
> > Is there a better tool for combining multiple dumps into a single pcap file?
> 
> No, but as I recall, it also lets you extract pieces of a pcap file by date, and usually that works out poorly. There is no better tool :-)

Wireshark's editcap can do that - but it does it with a sequential scan; tcpslice attempts to find the packets by probing the file - to quote the man page:

       tcpslice  cannot read its input from stdin, since it uses random‐access
       to rummage through its input files.

...

       tcpslice will not work properly on tcpdump files spanning more than one
       year; with files containing portions of packets whose  original  length
       was  more  than  65,535 bytes; nor with files containing fewer than two
       packets.  Such files result in the error message: ‘couldn’t find  final
       packet  in  file’.   These problems are due to the interpolation scheme
       used by tcpslice to greatly speed up its processing when  dealing  with
       large  trace  files.  Note that tcpslice can efficiently extract slices
       from the middle of trace files of any size,  and  can  also  work  with
       truncated  trace files (i.e., the final packet in the file is only par‐
       tially present, typically due to tcpdump being ungracefully killed).

so, while editcap doesn't have the "more than one year" issue, tcpslice may be faster than editcap on large files.

editcap also doesn't do merging (there's a separate mergecap utility for that), nor does it do session tracking to split the file into sessions:

       −s     Enable session tracking for  the  specified  types  which  is  a
              comma‐separated list of the following:

              tcp    track all TCP connections

              sip    track  SIP‐based VoIP calls, which may enable tracking of
                     TCP connections but only the ones that are related to SIP
                     calls.   This  feature  only  available  if  tcpslice was
                     linked against Aymeric Moizard’s  GNU  oSIP  library;  if
                     not,   install   the  latest  version  of  libosip2  from
                     https://www.gnu.org/software/osip/  and  recompile   tcp‐
                     slice.

              h323   track  H.323‐based  VoIP calls, which may enable tracking
                     of TCP connections but only the ones that are related  to
                     H.323 calls.  This feature only available if tcpslice was
                     linked against Objective Systems’ Open H.323 library  for
                     C;  if not, install the latest version of libooh323c from
                     https://sourceforge.net/projects/ooh323c/  and  recompile
                     tcpslice.

              Session  tracking  altogether  is only available if tcpslice was
              linked against a recent version (>1.20) of Rafal Wojtczuk’s Net‐
              work  Intrusion  Detection  System  library; if not, install the
              latest version of libnids  from  http://libnids.sourceforge.net/
              and recompile tcpslice.

**Comment by infrastation** [comment]
Thank you for waiting. tcpslice now incorporates a number of useful updates and builds on Linux, FreeBSD, macOS and Solaris. Let's see what work remains to be done before it can be called version 1.3.

Debian sid source package contains 8 patches to be applied to `tcpslice_1.2a3.orig.tar.gz` (original release from LBL):
```text
01_time.patch
02_largefile.patch
03_bpf_h.patch
04_zero_and_one_packet_captures.patch
05_fix_freed_memory_use.patch
06_fix-install.patch
07_fix-missing-include-stdlib.h.patch
08_hardening.patch
```
Could you please try the current master branch and confirm the package builds well? Also, if any of these patches still stand, would it be better to merge any of them them upstream?

**Comment by infrastation** [comment]
Let's wait for any other input for another week or so, and then just make a 1.3 release to introduce a common point of reference.

**Comment by infrastation** [comment]
To whom it may concern, meanwhile it became possible to retrieve exact copies of the 4 last LBL tcpslice releases from this repository as the following git tags (all on the "lbl" branch):
```text
tcpslice-1.1a3
tcpslice-1.2a1
tcpslice-1.2a2
tcpslice-1.2a3
```

**Comment by infrastation** [comment]
tcpslice 1.3 is now available as a git tag in this repository and as a [signed](http://www.tcpdump.org/release/tcpslice-1.3.tar.gz.sig) [tarball](http://www.tcpdump.org/release/tcpslice-1.3.tar.gz).

@brunonaibert, do you need any other help to proceed as originally intended?

**Comment by brunonaibert** [comment]
At the moment, only this issue.
Thanks @infrastation.
